encoding/ccf/sort.go

```diff
@@ -0,0 +1,207 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"bytes"
+
+	"github.com/onflow/cadence"
+)
+
+// bytewiseFieldSorter
+
+// bytewiseFieldSorter is used to sort fields by identifier.
+// NOTE: in order to avoid making a copy of fields for sorting,
+//   - create sorter by calling newBytewiseFieldIdentifierSorter()
+//   - sort by calling sort.Sort(sorter)
+//   - iterate sorted fields by
+//     for _, index := range sorter.indexes {
+//     // process sorted field at fields[index]
+//     }
+type bytewiseFieldSorter struct {
+	// NOTE: DON'T sort fields because it isn't a copy.
```

Comment:
Done in commit 9d172c996d5dc692ba465690a1744428a6806c3f here and in several other places.

Basically, `bytewiseFieldSorter` shouldn't sort fields **in place** because sorter doesn't make a copy of fields.  Instead it should sort `indexes` by field identifier.  Caller should iterate over sorted `indexes`.

```Go
sorter := newBytewiseFieldIdentifierSorter(fields)
sort.Sort(sorter)
for _, index := range sorter.indexes {
        // process sorted field at fields[index]
}
```

Reply:
When I asked this question initially, I was sort of getting the how, but was mostly wondering about the why. It seems like the answer is simply that we do not want to have the input being mutated, which makes sense 👍